### PR TITLE
Fix ERC20 token approval

### LINKED ISSUE #2010: Concurrent approvals will fail with RDN
https://github.com/raiden-network/light-client/blob/335cf61617c47854329c4664b428ab95d76329a9/raiden-ts/src/channels/epics.ts#L698-L702

the RDN smart contract only allows approval to change to and from zero. `0 -> x` and `x -> 0` where `x != 0` is okay. `x -> y` where both `x != 0` and `y != 0` will fail.

```solidity
    /// @notice Allows `_spender` to transfer `_value` tokens from `msg.sender` to any address.
    /// @dev Sets approved amount of tokens for spender. Returns success.
    /// @param _spender Address of allowed account.
    /// @param _value Number of approved tokens.
    /// @return Returns success of function call.
    function approve(address _spender, uint256 _value) public returns (bool) {
        require(_spender != 0x0);

        // To change the approve amount you first have to reduce the addresses`
        // allowance to zero by calling `approve(_spender, 0)` if it is not
        // already 0 to mitigate the race condition described here:
        // https://github.com/ethereum/EIPs/issues/20#issuecomment-263524729
        require(_value == 0 || allowed[msg.sender][_spender] == 0);

        allowed[msg.sender][_spender] = _value;
        Approval(msg.sender, _spender, _value);
        return true;
    }
```

This is also a problem in the Python implementation. Ideally concurrent approvals should be allowed since that can have a considerable impact on time for joining a network. (Perhaps not a big problem for the light client, if the expected number of channels is 1)

### PR DESCRIPTION
**Pull request by andrevmatos**
Fixes #2010 
Fixes UDC deposit part of #1908 

**Short description**
This makes 2 important changes:
- To approve before depositing (both for `depositChannel` and `depositToUDC`), if the current allowance is less than required to deposit but different than `0`, we must first set it to `0` then set the needed allowance; this is a requirement on secure ERC20 contracts like RDN
- To avoid the double transactions as above, we now default to approving `MaxUint256` (i.e. a very big number), trusting the `spender` (i.e. `TokenNetwork` contract and `UserDeposit` contract, respectively); this way, only an `approve` per token is needed, and the following deposits can happen without the need to mine `approve` again
  - If a user isn't confortable with the above, they can either set `config.minimumAllowance` to a small value, like `Zero`, which will make it default to the strictly requested `deposit` amount, and/or, on Metamask's prompt (per `approve` transaction), simply set the `allowance` to any desired amount (as long as it's greater than or equal the requested `deposit`, or else the `deposit` tx will fail)

Additionally, this also refactor the `depositToUDC` logic, which was on the public `Raiden` class, to its own epic, and interact with it through `udcDeposit` async actions.

**Definition of Done**

- [x] Steps to manually test the change have been documented
- [x] Acceptance criteria are met
- [ ] Change has been manually tested by the reviewer (dApp) 

**Steps to manually test the change (dApp)**

1. Test the default requested `approve` is now a very big number
2. Test editing allowance on Metamask and setting it to a value greater than the requested deposit will succeed (don't forget to also increase a little the `gasLimit`), but on the next deposit, it'll be required to first `approve(0)` and then the new value, which still defaults to a very big number

### CODE REVIEW

File: raiden-ts/src/services/epics.ts [hunk 1]
```diff
@@ -429,15 +442,133 @@ export const monitorUdcBalanceEpic = (
...
+      assert(deposited.add(deposit).eq(totalDeposit), [
+        ErrorCodes.UDC_DEPOSIT_OUTDATED,
+        { requested: totalDeposit.toString(), current: deposited.toString() },
+      ]);
+      assert(balance.gte(deposit), [
+        ErrorCodes.RDN_INSUFFICIENT_BALANCE,
+        { current: balance.toString(), required: deposit.toString() },
+      ]);
+
+      if (allowance.gte(deposit)) return of(true); // if allowance already enough
```
Comment (weilbith) [inline raiden-ts/src/services/epics.ts:480]: This caused me issues. It is always hard to follow these return statements. There is this fundamental thing in my head that return statement end a function and go back on the heap. Instead I had to remember myself that I'm in a mergeMap, jump to its end to see what happens (because the functions states to deposit, which didn't happen yet) and then come back to checkout the else case. Could we please try to export in-themselves closed maps so we have an easy to follow flow? So I could read: first we do the approve (if necessary) then the deposit. Then this statement in a dedicated function just for the allowance/approval would not cause these issues I explained.

File: raiden-ts/src/services/epics.ts [hunk 2]
```diff
@@ -429,15 +442,133 @@ export const monitorUdcBalanceEpic = (
...
+              txHash: receipt.transactionHash,
+              txBlock: receipt.blockNumber,
+              confirmed: undefined, // let confirmationEpic confirm this action
+            },
+            action.meta,
+          ),
+        ),
+        catchError((error) => of(udcDeposit.failure(error, action.meta))),
+      ),
+    ),
```
Comment (weilbith) [inline raiden-ts/src/services/epics.ts:569]: This function is pretty good. It mainly has a single layer of abstraction that could be followed. Most small details are hidden in dedicated functions that are named descriptive enough to understand what they do.

File: raiden-ts/src/raiden.ts [hunk 3]
```diff
@@ -1107,7 +1107,12 @@ export class Raiden {
    * @returns Promise to UDC remaining capacity
    */
   public async getUDCCapacity(): Promise<BigNumber> {
-    const balance = await this.deps.latest$.pipe(pluck('udcBalance'), first(isntNil)).toPromise();
+    const balance = await this.deps.latest$
+      .pipe(
+        pluck('udcBalance'),
+        first((balance) => !!balance && balance.lt(MaxUint256)),
```
[the remark this answers is not in the document]
Comment (andrevmatos) [reply raiden-ts/src/raiden.ts:1113]: No, `balance` is a `UInt<32>` object, and therefore, always truty

File: raiden-ts/src/raiden.ts [hunk 4]
```diff
@@ -1139,100 +1144,37 @@ export class Raiden {
...
    *    Set this to true if one wants to force sending the transaction with the subkey
-   * @param options.allowance - Value to approve on service token. Allows further deposits without
-   *    needing to approve again. Defaults to exact amount.
    * @returns transaction hash
    */
   public async depositToUDC(
     amount: BigNumberish,
     onChange?: OnChange<EventTypes, { txHash: string }>,
-    { subkey, allowance }: { subkey?: boolean; allowance?: BigNumberish } = {},
+    { subkey }: { subkey?: boolean } = {},
```
[the remark this answers is not in the document]
Comment (andrevmatos) [reply raiden-ts/src/raiden.ts:1152]: We're not yet constrained to backwards compatibility, and this was a very very edgy feature which I'm 100% sure no one of the early adopters used, so I took the freedom to remove it since it isn't relevant anymore with the default `minimumAllowance`, and the user can always edit the allowance on Metamask prompt anyway

File: raiden-ts/src/raiden.ts [hunk 5]
```diff
@@ -1139,100 +1144,37 @@ export class Raiden {
...
-        )
-      ).transactionHash as Hash;
-
-    onChange?.({ type: EventTypes.APPROVED, payload: { txHash: approveTxHash } });
+    const mined = asyncActionToPromise(udcDeposit, meta, this.action$).then((res) => res?.txHash);
+    this.store.dispatch(udcDeposit.request({ deposit, subkey }, meta));
 
-    const udcTotalDeposit = await userDepositContract.functions.total_deposit(this.address);
+    let txHash = await mined;
+    onChange?.({ type: EventTypes.DEPOSITED, payload: { txHash: txHash! } });
```
[the remark this answers is not in the document]
Comment (andrevmatos) [reply raiden-ts/src/raiden.ts:1170]: This is after mined, but the `confirmed` event may not happen due to a reorg removing it.

File: raiden-ts/src/services/epics.ts [hunk 6]
```diff
@@ -429,15 +442,133 @@ export const monitorUdcBalanceEpic = (
...
+ * @param deps.userDepositContract - UserDeposit contract instance
+ * @param deps.getTokenContract - Factory for Token contracts
+ * @param deps.address - Our address
+ * @param deps.log - Logger instance
+ * @param deps.signer - Signer
+ * @param deps.main - Main Signer/Address
+ * @param deps.config$ - Config observable
+ * @returns - Observable of udcDeposit.failure|udcDeposit.success actions
+ */
+export const udcDepositEpic = (
```
[the remark this answers is not in the document]
Comment (andrevmatos) [reply raiden-ts/src/services/epics.ts:530]: I agree, but this would be a deeper refactoring. Let's create a separate issue for this.

File: raiden-ts/src/services/epics.ts [hunk 7]
```diff
@@ -429,15 +442,133 @@ export const monitorUdcBalanceEpic = (
...
+): Observable<udcDeposit.success> =>
   latest$.pipe(
     pluckDistinct('state', 'blockNumber'),
     // it's seems ugly to call on each block, but UserDepositContract doesn't expose deposits as
     // events, and ethers actually do that to monitor token balances, so it's equivalent
     exhaustMap(() => userDepositContract.functions.effectiveBalance(address) as Promise<UInt<32>>),
-    distinctUntilChanged((x, y) => y.eq(x)),
-    map(udcDeposited),
+    withLatestFrom(latest$),
+    filter(([balance, { udcBalance }]) => !udcBalance.eq(balance)),
```
[the remark this answers is not in the document]
Comment (andrevmatos) [reply raiden-ts/src/services/epics.ts:452]: We don't. This isn't an event, just the current balance, but since the UDC doesn't have an event for this, it's harder to confirm the balance. An option would be to check `confirmationBlocks` in the past, and call it the confirmed balance. I'll implement.

File: raiden-ts/src/services/epics.ts [hunk 8]
```diff
@@ -429,15 +442,133 @@ export const monitorUdcBalanceEpic = (
...
+      tokenContract.functions.balanceOf(sender),
+      tokenContract.functions.allowance(sender, userDepositContract.address),
+      userDepositContract.functions.effectiveBalance(address),
+    ]),
+  ).pipe(
+    mergeMap(([balance, allowance, deposited]) => {
+      assert(deposited.add(deposit).eq(totalDeposit), [
+        ErrorCodes.UDC_DEPOSIT_OUTDATED,
+        { requested: totalDeposit.toString(), current: deposited.toString() },
+      ]);
```
[the remark this answers is not in the document]
Comment (andrevmatos) [reply raiden-ts/src/services/epics.ts:474]: This is a plain assertion which goes straight to `udcDeposit.failed` action, which then gets logged. No need to have its own log here.

File: raiden-ts/src/services/epics.ts [hunk 9]
```diff
@@ -429,15 +442,133 @@ export const monitorUdcBalanceEpic = (
...
+      // secure ERC20 tokens require changing allowance only from or to Zero
+      // see https://github.com/raiden-network/light-client/issues/2010
+      let resetAllowance$: Observable<true> = of(true);
+      if (!allowance.isZero())
+        resetAllowance$ = defer(() =>
+          tokenContract.functions.approve(userDepositContract.address, 0),
+        ).pipe(
+          assertTx('approve', ErrorCodes.RDN_APPROVE_TRANSACTION_FAILED, { log }),
+          mapTo(true),
+        );
```
[the remark this answers is not in the document]
Comment (andrevmatos) [reply raiden-ts/src/services/epics.ts:491]: Already thought about it. This is hardly splittable, since these return observable pipelines with different operations (`deposit` vs `setTotalDeposit`). These should be the only 2 types of contracts we approve for deposit, but I'll create an issue to evaluate if this is needed to be split/refactored.

File: raiden-ts/src/services/epics.ts (hunk 1 above)
Comment (andrevmatos) [reply raiden-ts/src/services/epics.ts:480]: Sorry if it may be hard for you to read these, but early `return` and early `throw` is essential in our workflow to _fail_ or _succeed_ early, in the sense of letting the bulk of operations in the end and with -1 indentation level as would be required. This is a pattern used thorough the codebase already, but we may discuss if it isn't reasonable and decide on a refactor.
I'll extract the `mergeMap`'s callback on these examples, but refactor patterns like these may be worth their own issue and proper discussion.

File: raiden-ts/src/services/epics.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (andrevmatos) [reply raiden-ts/src/services/epics.ts:530]: #2030

File: raiden-ts/src/services/epics.ts (hunk 1 above)
Comment (andrevmatos) [reply raiden-ts/src/services/epics.ts:480]: Done

File: raiden-ts/src/services/epics.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (andrevmatos) [reply raiden-ts/src/services/epics.ts:491]: I've split `approveIfNeeded$`, should be enough, what's left is small and specific enough to stay its own functions.

File: raiden-ts/src/raiden.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (weilbith) [reply raiden-ts/src/raiden.ts:1148]: Actually I meant that we still have an object `options: { subkey?: boolean }` instead of just `subkey = false`....

File: raiden-ts/src/services/epics.ts (hunk 1 above)
Comment (weilbith) [reply raiden-ts/src/services/epics.ts]: Thanks. I know these comments are hard. :see_no_evil:

**Comment by weilbith** [comment]
Thanks for your adjustment! :hugs:

File: raiden-ts/src/raiden.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (andrevmatos) [reply raiden-ts/src/raiden.ts:1148]: Oh, but `subkey = false` is a different behavior than `undefined`: The on-chain txs can be performed by the main key or the subkey, depending on the `config.subkey` value (defaults to using the main account when possible, but can be changed to `true` to use the subkey). If you pass `false` here, only for this operation, it'll force use the mainkey regardless of the `config.subkey`.

[the remark this answers is not in the document]
Comment (weilbith) [reply raiden-ts/src/raiden.ts:1148]: Well then `subkey?: boolean`. Doesn't matter.
Anyways. Out of focus now.
